server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java

```diff
@@ -1717,23 +1692,19 @@ private void upgradeTsFileResourceCallBack(TsFileResource tsFileResource) {
               resource.getEndTime(device))
       );
     }
-    insertLock.writeLock().lock();
-    tsFileManagement.writeLock();
-    try {
-      if (tsFileResource.isSeq()) {
-        tsFileManagement.addAll(upgradedResources, true);
-        upgradeSeqFileList.remove(tsFileResource);
-      } else {
-        tsFileManagement.addAll(upgradedResources, false);
-        upgradeUnseqFileList.remove(tsFileResource);
+    upgradeFileCount.getAndAdd(-1);
+    // load all upgraded resources in this sg to tsFileManagement
+    if (upgradeFileCount.get() == 0) {
```

Comment:
any side-effect that only moving finished files from upgrade list to the normal fileList  after all files are upgraded?